JoinColony/colonyCDapp
# Port Metatransactions, Update Transactions System

This PR ports over all the Metatransactions functionality (including the `broadcaster` service), along with all the other updates that were made to the transactions system in general.

Since it was not completed initially, we also have to logic to the Ganache Wallet Module to support message signing, typed data signing, and transactions sending

### Added / Changed
- [x] added broadcaster service (to dev environment)
- [x] Updated transaction sagas
- [x] Ported over Metatransaction sagas
- [x] Updated `ganacheModule` to support typed data signing
- [x] Fixed and set `ColonyManager` in the context
- [x] Ported over `GasStation` component
- [x] refactored `.env` variables _(basically this is the initial setup)_

### Removed
- [x] removed `util/debug`
- [x] removed `getProvider` saga _(as we will be relying on the user's wallet's provider from now on)_
- [x] removed `web3-utils` dependency
- [x] removed `promisify` utill 
- [x] removed `/inbox` and `/wallet` routes _(since we won't be porting over that part of the Dapp)_

Note that metatransactions, while _technically_ working, until there is support for them in `colonyJS` we can't actually use them.

### Testing

There's no easy way to test this, but in order to see a message and a transaction working, use the following:
- make sure colony / user data is seeded _(via the temp script)_
- make sure you are logged in with the first user _(`0xb77D57F4959eAfA0339424b83FcFaf9c15407461`)_
- Apply the following patch to your codebase _(just apply it, **don't commit it**)_
```diff
diff --git a/src/i18n/en.json b/src/i18n/en.json
index c969b8e..f3fde7a 100644
--- a/src/i18n/en.json
+++ b/src/i18n/en.json
@@ -410,5 +410,8 @@
     "permission.1": "Root",
 
     "tooltip.lockedToken": "This token is locked. Colony native tokens are locked and non-transferrable by default to avoid unwanted project token transfer outside of the colony.",
-    "tooltip.forceToggle" : "Toggle \"Force\" to bypass governance and perform this action immediately. This is only possible with the right permissions."
+    "tooltip.forceToggle" : "Toggle \"Force\" to bypass governance and perform this action immediately. This is only possible with the right permissions.",
+
+    "transaction.generic.title": "Test transaction",
+    "transaction.generic.description": "Test transaction description"
   }
diff --git a/src/redux/sagas/users/index.ts b/src/redux/sagas/users/index.ts
index 9497317..1f7b810 100644
--- a/src/redux/sagas/users/index.ts
+++ b/src/redux/sagas/users/index.ts
@@ -3,10 +3,12 @@ import {
   // fork,
   put,
   takeLatest,
+  fork,
 } from 'redux-saga/effects';
 // import { ClientType, TokenLockingClient } from '@colony/colony-js';
 // import { BigNumber } from 'ethers';
 
+import { ClientType } from '@colony/colony-js';
 import { ActionTypes } from '../../actionTypes';
 import {
   // Action,
@@ -14,13 +16,18 @@ import {
 } from '../../types/actions';
 import { ContextModule, removeContext } from '~context';
 import {
-  putError,
   // takeFrom,
   // createUserWithSecondAttempt,
   // getColonyManager
+  putError,
+  takeFrom,
 } from '../utils';
 import { clearLastWallet } from '~utils/autoLogin';
 
+import { signMessage } from '../messages';
+import { createTransaction, createTransactionChannels } from '../transactions';
+import { transactionReady, transactionPending } from '../../actionCreators';
+
 // import { transactionLoadRelated, transactionReady } from '../../actionCreators';
 // import {
 //   createTransactionChannels,
@@ -311,6 +318,40 @@ function* userLogout() {
 //   return null;
 // }
 
+function* testMessageSigning() {
+  yield signMessage('generic', 'hello world');
+}
+
+function* testTransaction() {
+  const batchKey = 'testTransaction';
+  const { testTxAction: testTx } = yield createTransactionChannels('someId', [
+    'testTxAction',
+  ]);
+  const createGroupTransaction = ({ id, index }, config) =>
+    fork(createTransaction, id, {
+      ...config,
+      group: {
+        key: batchKey,
+        id: 'someId',
+        index,
+        title: { id: 'transaction.generic.title' },
+        description: { id: 'transaction.generic.description' },
+      },
+    });
+  yield createGroupTransaction(testTx, {
+    context: ClientType.ColonyClient,
+    methodName: 'editColony',
+    identifier: '0x3a157280ca91bB49dAe3D1619C55Da7F9D4438c3',
+    params: ['fakeIPFSHash'],
+    ready: false,
+    title: { id: 'transaction.generic.title' },
+  });
+  yield takeFrom(testTx.channel, ActionTypes.TRANSACTION_CREATED);
+  yield put(transactionPending(testTx.id));
+  yield put(transactionReady(testTx.id));
+  yield takeFrom(testTx.channel, ActionTypes.TRANSACTION_SUCCEEDED);
+}
+
 export function* setupUsersSagas() {
   // yield takeLatest(ActionTypes.USER_AVATAR_REMOVE, userAvatarRemove);
   // yield takeLatest(ActionTypes.USER_AVATAR_UPLOAD, userAvatarUpload);
@@ -318,4 +359,6 @@ export function* setupUsersSagas() {
   // yield takeLatest(ActionTypes.USERNAME_CREATE, usernameCreate);
   // yield takeLatest(ActionTypes.USER_DEPOSIT_TOKEN, userDepositToken);
   // yield takeLatest(ActionTypes.USER_WITHDRAW_TOKEN, userWithdrawToken);
+  yield takeLatest('SIGN_MESSAGE', testMessageSigning);
+  yield takeLatest('SEND_TX', testTransaction);
 }

```
- Using Redux dev tools, dispatch the `SIGN_MESSAGE` action type to simulate signing a message
- Using Redux dev tools, dispatch the `SEND_TX` action type to simulate sending a transaction
![Screenshot from 2022-11-01 23-04-00](https://user-images.githubusercontent.com/1193222/199340905-237f9128-6573-4090-955f-7df4dd21ef8c.png)


Resolves #30
